### PR TITLE
Insert delay between shifted chars in send_string_with_delay

### PR DESCRIPTION
**Pull request by dead10ck**
## Description

When using `send_string_with_delay`, if the string contains capital letters, it gets sent to `send_char`, which does not insert the delay between pressing the modifier keys needed to type the shifted letter. This results in key codes being sent unshifted in some situations, such as ssh connections where the keyboard is too fast for the modifier taps to register.

This commit inesrts a delay between every modifier tap when they are detected.

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/send_string/send_string.h
Comment (fauxpark) [inline quantum/send_string/send_string.h:122]: This function is badly named and should really live in `wait.c`.

**Review by fauxpark — COMMENTED** [review]
I'm not sure if this should be done. `send_char` is meant to send a single character (not key event) in one operation - ideally a single report, but I don't know if this is what actually happens.

File: quantum/send_string/send_string.h
Comment (zvecr) [reply quantum/send_string/send_string.h:122]: It's also not necessary as all the wait implementations were aligned to allow both variable and constant arguments

Comment (dead10ck) [reply quantum/send_string/send_string.h:122]: Sorry, I don't understand. This doesn't change the behavior to sending more than one character. The existing code already sends several key events if the character is shifted.

The `send_string_with_delay` function already works with a delay time that is totally independent of the tap delay, so that e.g. `SEND_STRING_DELAY` can set an arbitrary delay between key presses. The problem is that the current code does not use a consistent delay between all the key presses because it doesn't use the input delay at all for ASCII chars. So e.g. if you do

```c
SEND_STRING_DELAY("tail -F file | grep \"hello\"", 20);
```

This will

1. wait 20ms for each of `tail -`
2. TAP_CODE_DELAY for shift down, `f`, shift up
3. 20ms for `file `
4. TAP_CODE_DELAY for shift down, `\`, shift up
5. 20ms for ` grep`
6. Etc, etc

>This function is badly named and should really live in wait.c

Ah, thanks for the pointer, I'm unfamiliar with the code base, so I'll take a look here.

Comment (dead10ck) [reply quantum/send_string/send_string.h:122]: Actually it's just calling `register_code` directly, so it may not even be waiting the tap delay in between the mods and the ASCII character?

File: quantum/send_string/send_string.c
Comment (fauxpark) [inline quantum/send_string/send_string.c:186]: This can be moved down to eliminate the duplicate calls above.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: quantum/send_string/send_string.c
Comment (dead10ck) [reply quantum/send_string/send_string.c:186]: I actually did that first, but that made it so that on ASCII characters, it would wait twice: once during `send_char_with_delay` and then again when it returned to the call site.

Comment (fauxpark) [reply quantum/send_string/send_string.c:186]: But that's in a separate branch. I'm not talking about where the loop was.

Comment (dead10ck) [reply quantum/send_string/send_string.c:186]: Ohh, you mean at the end of the inner block, right! Good catch. Will change.

File: quantum/send_string/send_string.c
Comment (fauxpark) [inline quantum/send_string/send_string.c:184]:
```suggestion
                wait_ms(ms);
            }

            wait_ms(interval);
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: quantum/send_string/send_string.c
Comment (dead10ck) [reply quantum/send_string/send_string.c:183]: Oh wait, actually, if we do this, then that means if there are any `SS_DELAY`s, it will wait that delay, and then the interval delay again. Is that okay? I guess that would make sense if the user did both an explicit delay as well as using `SEND_STRING_DELAY`

**Comment by fauxpark** [comment]
I tried to think of a way we could get the `send_char()` operation to use a single report by adding the modifiers to the keycode variable before giving it to `tap_code()` (using `tap_code16()` instead), but I don't think it's the best idea due to the way modifiers are encoded in the keycode. The low byte is used for the basic keycode, and the low 5 bits of the high byte for the modifiers; the top bit being the left/right flag therefore left shift and right alt can't be combined as it'd just be turned into right shift. This would probably work fine but maybe it'd be better to simply increase the size of the keycode lookup table to allow for modified keycodes. But that's something for another day...

**Review by fauxpark — APPROVED** [review]

File: quantum/send_string/send_string.c
Comment (fauxpark) [reply quantum/send_string/send_string.c:183]: If you want, you can rearrange this block to test for `SS_DELAY_CODE` first, then in an `else` do the rest, with the `wait_ms()` after.

Comment (dead10ck) [reply quantum/send_string/send_string.c:183]: The more I think about this, the more I think it is better to have it the way it is. `SEND_STRING_DELAY` advertises itself as inserting a delay at the end of every key. It seems weird to break that by adding "unless you give an explicit delay action".

Though I don't feel strongly about it, so I am happy to make that change if you feel this should be the behavior.

File: quantum/send_string/send_string.c
Comment (zvecr) [inline quantum/send_string/send_string.c:222]: @fauxpark  Would `tap_code_delay` be better here?
```suggestion
    tap_code_delay(keycode, interval);
```

Comment (dead10ck) [reply quantum/send_string/send_string.c]: Yeah that makes a lot of sense; done.

Comment (fauxpark) [reply quantum/send_string/send_string.c]: No, that controls the time between keydown and keyup of the keycode, not the time between keyup of the previous event and keydown of the next.

File: quantum/send_string/send_string.c
Comment (fauxpark) [inline quantum/send_string/send_string.c:150]:
```suggestion
    send_string_with_delay(string, 0);
```

File: quantum/send_string/send_string.c
Comment (fauxpark) [inline quantum/send_string/send_string.c:196]:
```suggestion
    send_char_with_delay(ascii_code, 0);
```

File: quantum/send_string/send_string.c
Comment (fauxpark) [inline quantum/send_string/send_string.c:222]:
```suggestion
    tap_code(keycode, interval);
```

File: quantum/send_string/send_string.h
Comment (fauxpark) [inline quantum/send_string/send_string.h:50]:
```suggestion
 * This function simply calls `send_string_with_delay(string, 0)`.
```

File: quantum/send_string/send_string.h
Comment (fauxpark) [inline quantum/send_string/send_string.h:62]:
```suggestion
 * \param interval The amount of time, in milliseconds, to wait before typing the next character.
```

File: quantum/send_string/send_string.h
Comment (fauxpark) [inline quantum/send_string/send_string.h:69]:
```suggestion
 * This function simply calls `send_char_with_delay(string, 0)`.
```

File: quantum/send_string/send_string.h
Comment (fauxpark) [inline quantum/send_string/send_string.h:79]:
```suggestion
 * \param interval The amount of time, in milliseconds, to wait in between key presses.
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: quantum/send_string/send_string.c
Comment (dead10ck) [reply quantum/send_string/send_string.c]: Hm, I was thinking it would make sense to use a consistent delay for every key event. Even without using the same delay for tapping, the input delay is what was used between key down for the mods and the ASCII character. If we want to distinguish, then should we use `TAP_CODE_DELAY` between key down on mods and the ASCII character, then the input delay after key ups? This seems like it could get confusing for users trying to figure out the right timing for `SEND_STRING _DELAY`.

In my opinion, it makes more sense to use a consistent delay for every key event. But I'll defer to your judgement.

Comment (zvecr) [reply quantum/send_string/send_string.c:222]: > In my opinion, it makes more sense to use a consistent delay for every key event.

This would be my expectation too.

Comment (fauxpark) [reply quantum/send_string/send_string.c:222]: Then we should rename `TAP_CODE_DELAY` to more effectively communicate this. Right now it kind of implies that it's the delay between keydown and keyup when you call `tap_code()`, because it is.

Comment (dead10ck) [reply quantum/send_string/send_string.c:222]: > Then we should rename `TAP_CODE_DELAY` to more effectively communicate this. Right now it kind of implies that it's the delay between keydown and keyup when you call `tap_code()`, because it is.

I mean I'm not too fussed about just using the tap code delay either way, so if this is a point of contention, I'm perfectly happy to just continue with previous semantics.

But with that said, I'm having trouble understanding what you're saying. In my view, `TAP_CODE_DELAY` is "delay between keydown and keyup when you call `tap_code()`", as you say. But the delay at issue here is the one in use in `SEND_STRING_DELAY`, and is, from my perspective, independent of `tap_code()`. The fact that it calls `tap_code()` internally is just an implementation detail. So to rename a long-standing configuration parameter because of its use as the default in an unrelated function seems puzzling to me.

**Comment by dead10ck** [comment]
```
Linking: .build/aeboards_ext65_rev3_via.elf                                                         [ERRORS]
 |
 | /avr/include/util/delay.h: In function 'tap_code':
 | /avr/include/util/delay.h:187:2: error: __builtin_avr_delay_cycles expects a compile time integer constant
 |   __builtin_avr_delay_cycles(__ticks_dc);
 |   ^
 | lto-wrapper: fatal error: avr-gcc returned 1 exit status
 | compilation terminated.
 | /bin/../lib/gcc/avr/8.3.0/../../../../avr/bin/ld: error: lto-wrapper failed
 | collect2: error: ld returned 1 exit status
 |
gmake[1]: *** [builddefs/common_rules.mk:267: .build/aeboards_ext65_rev3_via.elf] Error 1
```

Wait, I'm confused, I thought `wait_ms` no longer needed a constant argument? Why does this error get thrown from `tap_code()` but not in all the other places it was changed?

File: quantum/send_string/send_string.c
Comment (fauxpark) [reply quantum/send_string/send_string.c:222]: > But with that said, I'm having trouble understanding what you're saying.

You're suggesting we use `TAP_CODE_DELAY` as the default value for *every* delay that sendstring does, which includes the ones you're introducing, between the key events. Not just in the `tap_code()`. To me this muddies the meaning of the define's name, so if we want this value to be applied to more than just `tap_code()`, it should probably be renamed to be more generic. I'm not saying that should be done in *this* PR, just noting a potential point of confusion for users.

**Comment by dead10ck** [comment]
Curiously, rebased on top of the latest `development`, now the QMK CI Build fails with

```
platforms/chibios/drivers/i2c_master.c: In function 'i2c_stop':
./keyboards/dztech/tofu/jr/v1/config.h:20:20: error: 'I2CD2' undeclared (first use in this function); did you mean 'I2CD1'?
 #define I2C_DRIVER I2CD2
                    ^~~~~
platforms/chibios/drivers/i2c_master.c:216:14: note: in expansion of macro 'I2C_DRIVER'
     i2cStop(&I2C_DRIVER);
              ^~~~~~~~~~
 [ERRORS]
 |
 |
 |
gmake[1]: *** [builddefs/common_rules.mk:360: .build/obj_dztech_tofu_jr_v1_via/i2c_master.o] Error 1

```

which doesn't look like it has anything to do with my change.

Is there anything else blocking this PR? Are we just waiting for another review?

**Comment by dead10ck** [comment]
Bumping this again. Are there any further changes anyone would like to see?

**Comment by dead10ck** [comment]
Friendly ping again

File: quantum/action.c
Comment (drashna) [inline quantum/action.c:973]: I believe that this needs to be reverted, as wait_ms may need an actual set amount at compile time on some platforms. Hence the for loop here.

Comment (sigprof) [reply quantum/action.c:973]: That should have been fixed for AVR long ago (and other platforms never had that problem):
```c
#define wait_ms(ms)                             \
    do {                                        \
        if (__builtin_constant_p(ms)) {         \
            _delay_ms(ms);                      \
        } else {                                \
            for (uint16_t i = ms; i > 0; i--) { \
                _delay_ms(1);                   \
            }                                   \
        }                                       \
    } while (0)
```

**Comment by dead10ck** [comment]
Is anything else needed for this PR?

**Comment by dead10ck** [comment]
Still waiting on feedback about this PR; it's not stale as far as I know.

**Review by drashna — APPROVED** [review]

**Comment by dead10ck** [comment]
Still not stale.

**Comment by dead10ck** [comment]
There are some CI failures about the resulting binary size being too large. It seems unlikely to me that my change is related at all.

**Comment by dead10ck** [comment]
Anything I can do for this PR?

**Review by tzarc — APPROVED** [review]

**Pull request merged**
